docs/DesignDocs/NotesOnAttestationAPI.md

```diff
@@ -0,0 +1,366 @@
+
+In this section, we focus on the discussion of the options for initialization
+of attester plugins. Same options are available for initialization of
+enclave-side verifier plugins.
+
+1. Every enclave-side plugin library implements the same OE SDK API
+`oe_attester_initialize()`.
+
+    With this option, a build of an application enclave project can only be linked
+    with a single attestation plugin library, to avoid function name collision.
+
+2. Build-time macro to convert from TEE-agnostic API to TEE-specific
+implementation.
+
+    With this option, application enclaves still call TEE-agnostic OE SDK API
+    `oe_attester_initialize()`.
+    But the OE SDK provides a macro definition of this function calling
+    the TEE- and library-specific initialization functions of linked
+    enclave-side plugin libraries.
+
+    This option has the potential to support linking with multiple
+    enclave-side plugin libraries in a single application enclave build
+    configuration.
+    The OE SDK build environment is required to provide a proper
+    `oe_attester_initialize()` macro definition for every configuration
+    it supports.
+
+3. Build-time manifest to list plugin libraries and their
+initialization entry points
+
+    With this option, every plugin library has its own initialization
+    entry point with globally unique name and well-defined prototype.
+    To build and sign an attester application enclave, a manifest / metadata
+    (.conf file) is created that contains the list of linked enclave-side
+    plugin libraries and their entry points. This manifest is compiled and
+    signed as part of the attester enclave image.
+
+    During run-time, upon load of an application enclave, its OE SDK framework
+    reads its embedded manifest, retrieves the addresses of all the listed
+    initialization function entry points, and invokes them in the order listed.
+
+    Note: this option requires the definition of the manifest format and the
+    new capability for the OE SDK build and signing tools to support processing of
+    manifest files.
+
+4. Run-time enclave-side plugin libraries registration of their initialization
+functions with the OE SDK framework.
+
+    The design document
+    [OpenEnclave Modules](https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md)
+    describes how an enclave-side module can register its initialization
+    function for invocation on first entry to the enclave (upon enclave load).
+
+    With this option, the implementation of the OE SDK attestation framework
+    provides a hook for enclave-side plugin libraries to register their
+    initialization functions upon enclave load.
+
+    If ordering between the initialization functions of multiple libraries is
+    required, it can be supported by assigning different priorities to the
+    initialization functions of these libraries when they are registered
+    upon load.
```

Comment:
If I understand right, this means that if you're writing a new library for which ordering would matter, the priority is hard coded right?  So the priorities must be known of any other libraries that you might care about ordering before or after.  If you care about one that isn't part of OE SDK and isn't open source, is there an API or log to figure out the priority?   (This is one reason I don't like this option, which applies to https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md as well I think.)   But as this document is not proposing something new, the text is ok, and we can track this as part of issue #2903

Reply:
Added note on lack of scalability.

Yes, agree with your comments about ordering.